service/history/ndc/workflow_resetter.go

```diff
@@ -866,6 +869,26 @@ func reapplyEvents(
 				return reappliedEvents, err
 			}
 			reappliedEvents = append(reappliedEvents, event)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CANCEL_REQUESTED:
+			if excludeCancelRequest || isDuplicate(event) {
+				continue
+			}
+			attr := event.GetWorkflowExecutionCancelRequestedEventAttributes()
+			request := &historyservice.RequestCancelWorkflowExecutionRequest{
+				CancelRequest: &workflowservice.RequestCancelWorkflowExecutionRequest{
+					Reason:   attr.GetCause(),
+					Identity: attr.GetIdentity(),
+					Links:    event.Links,
+				},
+				ExternalInitiatedEventId:  attr.GetExternalInitiatedEventId(),
+				ExternalWorkflowExecution: attr.GetExternalWorkflowExecution(),
+			}
+			if _, err := mutableState.AddWorkflowExecutionCancelRequestedEvent(
```

Comment:
We need to check if cancellation is already requested or not before adding the event, otherwise an error will be returned. 

The deduplication logic in `isDuplicate` is only best effort as the `appliedEvents` field is not persisted into DB. Actually even if it's persisted, there could a case where two different runs for the same workflowID both trying to reapply a canceled event to a third target run I think.

Reply:
agree. The dedup logic is a little different than other events. The cancel request only allow once per wf run. You can check the mutable status to check if the cancel is requested.